bin/propolis-utils/src/bin/cpuid-gen.rs

```diff
@@ -261,15 +262,116 @@ fn print_toml(results: &BTreeMap<CpuidKey, Cpuid>) {
+        match results.get(&CpuidKey::Leaf(0)) {
+            None => {
+                eprintln!("no result for leaf 0, selecting default vendor");
+                CpuidVendor::Amd
+            }
+            Some(values) => {
+                let values = CpuidValues {
+                    eax: values.eax,
+                    ebx: values.ebx,
+                    ecx: values.ecx,
+                    edx: values.edx,
+                };
+
+                match CpuidVendor::try_from(values) {
+                    Err(_) => {
+                        eprintln!(
+                            "vendor in leaf 0 values ({values:?}) \
+                        not recognized, selecting default vendor"
+                        );
+                        CpuidVendor::Amd
+                    }
+                    Ok(v) => v,
+                }
+            }
+        }
+    };
+
+    let cpuid = Cpuid {
+        entries: results
+            .iter()
+            .filter_map(|(key, value)| {
+                // propolis-server will complain if it receives a CPUID entry
+                // with no subleaf and a CPUID entry with a subleaf for the same
+                // leaf. Avoid emitting these by filtering out subleaf-less
+                // entries for any leaf that has an entry with subleaf data.
+                let (leaf, subleaf) = match key {
+                    CpuidKey::Leaf(leaf) => {
+                        if results.keys().any(|k| {
+                            matches!(k, CpuidKey::SubLeaf(l, _) if *l == *leaf)
+                        }) {
+                            return None;
+                        }
+
+                        (*leaf, None)
+                    }
+                    CpuidKey::SubLeaf(leaf, subleaf) => (*leaf, Some(*subleaf)),
+                };
+
+                Some(CpuidEntry {
+                    leaf,
+                    subleaf,
+                    eax: value.eax,
+                    ebx: value.ebx,
+                    ecx: value.ecx,
+                    edx: value.edx,
+                })
+            })
+            .collect(),
+        vendor,
+    };
+
+    println!("{}", serde_json::to_string_pretty(&cpuid).unwrap());
+}
+
+#[derive(Default, Clone, Copy, Debug)]
+enum OutputFormat {
+    /// Print a human-readable plain-text representation.
+    #[default]
+    Text,
+
+    /// Print TOML suitable for use in a propolis-standalone config file.
+    Toml,
+
+    /// Print JSON suitable for inclusion in a propolis-server instance spec.
+    Json,
+}
+
+impl FromStr for OutputFormat {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(match s {
+            "text" => Self::Text,
+            "toml" => Self::Toml,
+            "json" => Self::Json,
+            _ => {
+                return Err(
+                    "invalid output format, must be text, toml, or json",
+                )
+            }
+        })
+    }
+}
```

Comment:
nitpicky: If we just derived Clap's `ValueEnum` trait for this (see: https://docs.rs/clap/latest/clap/_derive/_tutorial/chapter_3/index.html#enumerated-values), it would do this for us, and also show the possible values in the help text, which is kinda nice...

Reply:
This is much, much nicer. Thanks for pointing it out! Done in 4aaff74.